src/cudamatrix/cu-vector.cc

```diff
@@ -711,6 +711,47 @@ void CuVectorBase<Real>::MulElements(const CuVectorBase<Real> &v) {
 
 template<>
 template<>
+void CuVectorBase<double>::CopyFromVec(const CuVectorBase<double> &src) {
+  KALDI_ASSERT(src.Dim() == dim_);
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    if (dim_ == 0) return;
+    Timer tim;
+    int dimBlock(CU2DBLOCK);
+    int dimGrid(n_blocks(dim_, CU2DBLOCK));
+    cuda_copy_from_vec(dimGrid, dimBlock, data_, src.data_, dim_);
+    CU_SAFE_CALL(cudaGetLastError());
+    CuDevice::Instantiate().AccuProfile(__func__, tim.Elapsed());
+  } else
+#endif
+  {
+    Vec().CopyFromVec(src.Vec());
+  }
+}
+
+template<>
+template<>
+void CuVectorBase<float>::CopyFromVec(const CuVectorBase<float> &src) {
```

Comment:
The whole idea behind this was to extend the cublas_copy template so that
when called for different types it would still work- it would have to
follow the same interface as the existing cublas_copy function.  The idea
is that that would then be called (with stride 1) from CopyFromVec; and
there are may be other situations where it would be useful too, when we
need strides.  We could then get rid of the functions copy_from_vec_fd etc.
Dan


On Wed, Apr 6, 2016 at 1:00 PM, pegahgh <notifications@github.com> wrote:

> In src/cudamatrix/cu-vector.cc
> <https://github.com/kaldi-asr/kaldi/pull/663#discussion_r58742474>:
>
> > +    Timer tim;
> > +    int dimBlock(CU2DBLOCK);
> > +    int dimGrid(n_blocks(dim_, CU2DBLOCK));
> > +    cuda_copy_from_vec(dimGrid, dimBlock, data_, src.data_, dim_);
> > +    CU_SAFE_CALL(cudaGetLastError());
> > +    CuDevice::Instantiate().AccuProfile(__func__, tim.Elapsed());
> > +  } else
> > +#endif
> > +  {
> > +    Vec().CopyFromVec(src.Vec());
> > +  }
> > +}
> > +
> > +template<>
> > +template<>
> > +void CuVectorBase<float>::CopyFromVec(const CuVectorBase<float> &src) {
>
> I think you don't need this float to float and double to double version of
> CopyFromVec and the vectors with same type is handled in the main template
> version of function!
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/663/files/059676582df8a41a59bcd4430483de45ba641194#r58742474>
>